Layout milestone notifications should be dispatched asynchronously

jelee53 opened the pull request:
#### bf24e9a12c1769d4eaaec4c8277addfc0f02d180
<pre>
Layout milestone notifications should be dispatched asynchronously
<a href="https://bugs.webkit.org/show_bug.cgi?id=303958">https://bugs.webkit.org/show_bug.cgi?id=303958</a>
<a href="https://rdar.apple.com/153254633">rdar://153254633</a>

Reviewed by Ryosuke Niwa.

When a layout milestone notification fires, the layout may not be completed but client
callbacks to layout milestone notifications can trigger Javascript execution. This results
in javascript execution failing (and consequently crashing) to run during the
layout process because it is correctly not permitted to do so. We delay firing layout
milestone notifications by dispatching them asychronously in order to prevent this.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/AutoLayoutIntegration.mm

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didReachLayoutMilestone):
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::willDispatchDidReachLayoutMilestone):
(WebFrameLoaderClient::dispatchDidReachLayoutMilestone):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AutoLayoutIntegration.mm:
(TEST(WebKit, AutoLayoutBatchesUpdatesWhenInvalidatingIntrinsicContentSize)):

Canonical link: <a href="https://commits.webkit.org/304572@main">https://commits.webkit.org/304572@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/abde782f70f30407f188a4b0319934f35a45ad3e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135956 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8321 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47239 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143665 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/88172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7c4b4a3d-a95e-4ecc-8ef4-456b5cdf0e42) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137825 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8983 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8166 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103900 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/88172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138902 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6502 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121843 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84777 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3fdb33ad-d0fb-4c68-b862-a7cc965508d8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6198 "Passed tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/3848 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4268 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115464 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40034 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146415 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8004 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40602 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112250 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8025 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6719 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112643 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6113 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118144 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20944 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8052 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36219 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7777 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71607 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7998 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7858 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -3081,7 +3082,9 @@ void FrameLoader::didReachLayoutMilestone(OptionSet<LayoutMilestone> milestones)
     ASSERT(m_frame->isMainFrame());
 
-    m_client->dispatchDidReachLayoutMilestone(milestones);
+    m_frame->document()->checkedEventLoop()->queueTask(TaskSource::Reporting, [this, protectedThis = Ref { *this }, milestones] {
```
Oh, I don't think we should be using `TaskSource::Reporting` for this. Probably the most appropriate one is `InternalAsyncTask`.

rniwa commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
```diff
@@ -873,6 +873,10 @@ void WebLocalFrameLoaderClient::dispatchDidReachLayoutMilestone(OptionSet<WebCor
 
     if (milestones & WebCore::LayoutMilestone::DidFirstVisuallyNonEmptyLayout) {
+
+        if (!m_didCompletePageTransition)
```
Why is this needed here??

jelee53 replied:
I sync'd with Ryosuke offline but basically, all the failing tests are failing on this assertion: `ASSERTION FAILED: !m_frame->isMainFrame() || webPage->corePage()->settings().suppressesIncrementalRendering() || m_didCompletePageTransition` 

`m_didCompletePageTransition` is false when it should be true. Since we asynchronously schedule the notification now, this notification is fired after `WebLocalFrameLoaderClient::provisionalLoadStarted()` starts which in turn sets `m_didCompletePageTransition` to false. 

So I just was adding this boolean to see if preventing layout-milestone-notifications during provisional loads would fix all the issues.

jelee53 replied:
I chatted with Ryosuke about how to better fix this and he was saying that we should only dispatch the layout milestone notification if the navigation ID from when the task is original queue'd matches the ID at dispatch time.

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/AutoLayoutIntegration.mm:
```diff
@@ -246,16 +246,17 @@ - (dispatch_block_t)invalidatedIntrinsicContentSizeBlock
     [webView loadHTMLString:@"<body style='margin: 0; height: 400px;'></body>" baseURL:nil];
     TestWebKitAPI::Util::run(&didInvalidateIntrinsicContentSize);
-    TestWebKitAPI::Util::run(&didFinishNavigation);
+    TestWebKitAPI::Util::run(&didFirstLayout);
```
Why is this change needed? Is the test relying on some side effect of calling the layout milestone?

jelee53 replied:
(Sync'd offline with Ryosuke) 

But just for posterity / in case anyone read this later, this test relied on the layout milestone notification finishing before navigation. `invalidateIntrinsicContentSize` occurs for the initial page load and once navigation finishes, there is a block added to `invalidateIntrinsicContentSize` and we make sure that `invalidateIntrinsicContentSize ` isn't called again. 

With our changes, the layout milestone notification fires after navigation finishes. So this block incorrectly tracks that `invalidateIntrinsicContentSize` of the initial page load. 

So I changed this so that we wait until after the layout finishes (rather than the navigation finishing) to add the block.

jelee53 committed:
Layout milestone notifications should be dispatched asynchronously
https://bugs.webkit.org/show_bug.cgi?id=303958
rdar://153254633

Reviewed by Ryosuke Niwa.

When a layout milestone notification fires, the layout may not be completed but client
callbacks to layout milestone notifications can trigger Javascript execution. This results
in javascript execution failing (and consequently crashing) to run during the
layout process because it is correctly not permitted to do so. We delay firing layout
milestone notifications by dispatching them asychronously in order to prevent this.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/AutoLayoutIntegration.mm

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didReachLayoutMilestone):
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::willDispatchDidReachLayoutMilestone):
(WebFrameLoaderClient::dispatchDidReachLayoutMilestone):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AutoLayoutIntegration.mm:
(TEST(WebKit, AutoLayoutBatchesUpdatesWhenInvalidatingIntrinsicContentSize)):

Canonical link: https://commits.webkit.org/304572@main
